# MuPDF f030eda1e472268667805f438e38cee8f1da61f8.
# Scripted buttons use the same name-valued group/appearance owner as clicks.
diff --git a/source/pdf/pdf-form.c b/source/pdf/pdf-form.c
--- a/source/pdf/pdf-form.c
+++ b/source/pdf/pdf-form.c
@@ -158,7 +158,6 @@
 	if (old_text && !strcmp(old_text, text))
 		return;
 
-	// TODO: if field is a checkbox, V should be a name and not a string!
 	pdf_dict_put_text_string(ctx, obj, PDF_NAME(V), text);
 
 	pdf_field_mark_dirty(ctx, obj);
@@ -723,34 +722,24 @@
 	return 1;
 }
 
-static void update_checkbox_selector(fz_context *ctx, pdf_document *doc, pdf_obj *field, const char *val)
-{
-	pdf_obj *kids = pdf_dict_get(ctx, field, PDF_NAME(Kids));
-
-	if (kids)
-	{
-		int i, n = pdf_array_len(ctx, kids);
-
-		for (i = 0; i < n; i++)
-			update_checkbox_selector(ctx, doc, pdf_array_get(ctx, kids, i), val);
-	}
-	else
-	{
-		pdf_obj *n = pdf_dict_getp(ctx, field, "AP/N");
-		pdf_obj *oval;
-
-		if (pdf_dict_gets(ctx, n, val))
-			oval = pdf_new_name(ctx, val);
-		else
-			oval = PDF_NAME(Off);
-		pdf_dict_put_drop(ctx, field, PDF_NAME(AS), oval);
-	}
-}
-
 static int set_checkbox_value(fz_context *ctx, pdf_document *doc, pdf_obj *field, const char *val)
 {
-	update_checkbox_selector(ctx, doc, field, val);
-	update_field_value(ctx, doc, field, val);
+	pdf_obj *grp = find_head_of_field_group(ctx, field);
+	pdf_obj *name = pdf_new_name(ctx, val ? val : "");
+	if (!grp)
+		grp = field;
+
+	fz_try(ctx)
+	{
+		/* Buttons select existing appearances; they do not regenerate them. */
+		if (!pdf_name_eq(ctx, pdf_dict_get(ctx, grp, PDF_NAME(V)), name))
+			pdf_dict_put(ctx, grp, PDF_NAME(V), name);
+		set_check_grp(ctx, doc, grp, name);
+	}
+	fz_always(ctx)
+		pdf_drop_obj(ctx, name);
+	fz_catch(ctx)
+		fz_rethrow(ctx);
 	return 1;
 }
 
